src/Illuminate/Support/Url.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+namespace Illuminate\Support;
+
+use Illuminate\Contracts\Support\Arrayable;
+use Stringable;
+
+class Url implements Arrayable, Stringable
+{
+    /**
+     * Constructor.
+     */
+    public function __construct(
+        public ?string $scheme = null,
+        public ?string $host = null,
+        public ?int $port = null,
+        public ?string $user = null,
+        public ?string $pass = null,
+        public ?string $path = null,
+        public ?UrlQueryParameters $query = null,
+        public ?string $fragment = null,
+    ) {
+        $this->query ??= new UrlQueryParameters;
+    }
+
+    /**
+     * Parse a URL string into a URL object.
+     *
+     * @param  string  $url
+     * @return static
+     */
+    public static function parse(string $url): static
+    {
+        $components = parse_url($url) ?: [];
+
+        if (isset($components['query'])) {
+            $components['query'] = static::query($components['query']);
+        }
+
+        return new static(...$components);
+    }
```

Comment:
Should this throw an exception if it is unable to parse the url into something meaningful, i.e., if `parse_url` fails and returns `false`?

Reply:
I think this is a good idea. Will change and mark resolved once pushed.

Appreciate the reviews!

Reply:
Done! 👍